src/node_wasm_streaming.cc

```diff
@@ -0,0 +1,377 @@
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#include "node_buffer.h"
+#include "node_errors.h"
+#include "node_external_reference.h"
+#include "node_wasm_streaming.h"
+
+#include "env-inl.h"
+
+namespace node {
+
+using errors::TryCatchScope;
+using v8::ArrayBuffer;
+using v8::Context;
+using v8::DataView;
+using v8::Function;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::Global;
+using v8::HandleScope;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::ObjectTemplate;
+using v8::Value;
+
+namespace WasmStreaming {
+namespace {
+
+class WasmStreaming final : public BaseObject {
+ public:
+  WasmStreaming(Environment* env,
+                Local<Object> wrap,
+                std::shared_ptr<v8::WasmStreaming> wasm_streaming);
+
+  static void OnBytesReceived(const FunctionCallbackInfo<Value>& args);
+  static void Finish(const FunctionCallbackInfo<Value>& args);
+  static void Abort(const FunctionCallbackInfo<Value>& args);
+  static void SetCompiledModuleBytes(const FunctionCallbackInfo<Value>& args);
+  static void SetUrl(const FunctionCallbackInfo<Value>& args);
+
+  SET_NO_MEMORY_INFO()
+  SET_MEMORY_INFO_NAME(WasmStreaming)
+  SET_SELF_SIZE(WasmStreaming)
+
+ private:
+  class OnModuleCompiledCallbackReg final : public v8::Task {
+   public:
+    OnModuleCompiledCallbackReg(Environment* env, Local<Function> callback);
+    ~OnModuleCompiledCallbackReg();
+    static void Invoke(std::unique_ptr<OnModuleCompiledCallbackReg> reg,
+                       v8::OwnedBuffer compiledModuleData);
+    void Run() override;
+
+   private:
+    std::shared_ptr<v8::TaskRunner> runner_;
+    v8::OwnedBuffer compiledModuleData_;
+    Environment* env_;
+    Global<Function> callback_;
+    async_context async_context_;
+  };
+
+  class Client final : public v8::WasmStreaming::Client {
+   public:
+    ~Client() override {
+      // Must support being deleted from any thread.
+      //
+      // Invoke with dummy zero-sized buffer if no compiled module was
+      // produced.  Could happend if Module was GC-ed before top tier
+      // compilation completes.  Takes care of reg_; has to be destroyed
+      // on the main thread.
+      OnModuleCompiledCallbackReg::Invoke(std::move(reg_), {});
+    }
+    // Passes the fully compiled module to the client.
+    void OnModuleCompiled(v8::CompiledWasmModule module) override {
+      OnModuleCompiledCallbackReg::Invoke(std::move(reg_), module.Serialize());
+    }
+    void InstallCallback(std::unique_ptr<OnModuleCompiledCallbackReg> reg) {
+      reg_ = std::move(reg);
+    }
+   private:
+    std::unique_ptr<OnModuleCompiledCallbackReg> reg_;
+  };
+
+ private:
+  std::shared_ptr<v8::WasmStreaming> wasm_streaming_;
+  std::shared_ptr<Client> client_;
+  Global<Value> compiled_module_bytes_;
```

Comment:
This is just me doing a first read of this, but I’ll mention that often it’s nicer to use internal fields to store JS values than `Global`s, because it’s really easy to create memory leaks when using the latter (if a property on the value refers back to the object which owns the `Global`) – you can look for `enum InternalFields` in the source code to see how we usually do this